### PR TITLE
Fix error when saving plots with normalisation

### LINKED ISSUE #30315: Colourfill plot with bin width normalisation causes project save to fail
### Expected behavior
If project recovery runs while a colourfill plot with bin width normalisation set, it should save the interface successfully.

### Actual behavior
It does not. `Unknown error occurred. Full detail: Object of type Normalize is not JSON serializable`

Edit: this also seems to happen even after closing the plot

### Steps to reproduce the behavior
Plot a colorfill plot, right click -> normalisation -> bin width, wait for project recovery to trigger and check the logger

### Platforms affected
Confirmed on windows, assumed on all

### PR DESCRIPTION
**Pull request by clayton-tom**
**Description of work.**

In the dict that is encoded for plots, normalise settings were saved in place as Normalise objects which cannot be serialised. This adds a conversion to dict and a Normalise construction from the dict to the encoder and decoder respectively.

**To test:**
1. Load a workspace, plot a colourfill and a single spectra plot
2. Set one or the other to Normalisation -> Bin width and leave the other off
3. Wait for project save to kick in or save the project manually, make sure no errors occur for saving the plots
4. Segfault and restore, or restart mantid and load project
5. Check that the plots have survived as they were with the correct normalisation options

Fixes #30315  

*This does not require release notes* because **it fixes a small bug discovered in manual testing**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
Your change does correctly save and load in the normalisation but it doesn't seem to update the normalization status in the menu and the axes label. The plotted data in the screenshot were saved with no normalisation. When I set normalisation back to None the counts/colorscale doesn't change and the label is corrected (i.e. becomes just counts rather than counts/mus). 
image

**Comment by RichardWaiteSTFC** [comment]
I'm still getting the same problem for colorfill plots, though it seems to be fixed for line plots (maybe it was never broken for line plots...?).  Can you reproduce it? Also is it possible to add some tests for this as well, it might not be but I notice there is a `test_plotssaver.py` file.

**Comment by clayton-tom** [comment]
@RichardWaiteSTFC sorry for the delay, finally got there after making it much more complicated that it needed to be

File: qt/python/mantidqt/project/plotsloader.py [hunk 1]
```diff
@@ -124,7 +122,11 @@ def workspace_plot_func(self, workspace, axes, fig, creation_arg):
         func = function_dict[function_to_call]
         # Plotting is done via an Axes object unless a colorbar needs to be added
         if function_to_call in ["imshow", "pcolormesh"]:
-            func([workspace], fig)
+            if creation_arg['normalize_by_bin_width']:
+                is_norm = creation_arg.pop('normalize_by_bin_width')
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/project/plotsloader.py:126]: Just checking why you need to pop the normalise argument, why can't you just do 
`func([workspace], fig, normalize_by_bin_width=creation_arg['normalize_by_bin_width'])`

File: qt/python/mantidqt/plotting/functions.py [hunk 2]
```diff
@@ -271,19 +273,23 @@ def pcolormesh(workspaces, fig=None):
...
     ax.clear()
     ax.set_title(ws.name())
     scale = _get_colorbar_scale()
     if use_imshow(ws):
         pcm = ax.imshow(ws, cmap=ConfigService.getString("plots.images.Colormap"), aspect='auto', origin='lower',
-                        norm=scale())
+                        norm=scale(), normalize_by_bin_width=normalize_by_bin_width)
+        # remove normalize_by_bin_width from cargs if present so that this can be toggled in future
+        for cargs in pcm.axes.creation_args:
+            _ = cargs.pop('normalize_by_bin_width')
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/plotting/functions.py:292]: If you don't need the return argument you don't have to assign anything, you can just do 
`cargs.pop('normalize_by_bin_width')`

File: qt/python/mantidqt/project/plotssaver.py [hunk 3]
```diff
@@ -14,9 +15,9 @@
 from mantid.plots.legend import LegendProperties
 
 try:
-    from matplotlib.colors import to_hex
+    from matplotlib.colors import to_hex, Normalize
 except ImportError:
-    from matplotlib.colors import colorConverter, rgb2hex
+    from matplotlib.colors import colorConverter, rgb2hex, Normalize
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/project/plotssaver.py:20]: Is this a python 2/3 thing we can get rid of?

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks good! The colorfill and line plot normalisation is correctly saved and loaded, independent of the default normalisation settings.

**Pull request merged**